File: c/simple_menu.c
#include <stdbool.h>
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <stdlib.h>

#define MAX_COMMAND_CHARACTERS 6
#define MAX_COMMAND_PARAMETERS 5
#define FIELD_SEPERATOR        ","
#define MAX_INPUT_CHARACTERS   100


void print_help_menu() {
    printf("This is where a description of the commands would be\n");
}

void print_button_command() {
    // Simulate if the button is pressed or not
    int pressed = rand() % 2;
    printf("Button status: %s\n", pressed ? "pressed" : "Not pressed");
}

void print_led_command(char *parameters[], __uint8_t args) {
    
    if(args == 0) {
        printf("Led command wrong parameters\n");
        return;
    }

    if (strncmp("on", parameters[0], 5) == 0) {    
       printf("Led is on\n");
    } else if (strncmp("off", parameters[0], 5) == 0) {
       printf("Led is off\n");
    } else if (strncmp("blink", parameters[0], 5) == 0) {
        if(args != 2) {
            printf("Led blinking command wrong parameters\n");
        }
        printf("Led is blinking with %s Hz\n", parameters[1]);
    } else {
       printf("Led command wrong parameters\n");
    }
    
} 

bool process_buffer(char **command, char *parameters[], __uint8_t *args, char *buffer) {

    char *rest;

    if (buffer == NULL || strlen(buffer) == 0) {
        return true;
    }

    // First token is the command string
    *command = strtok_r(buffer, FIELD_SEPERATOR, &rest);
    if (strlen(*command) == 0) {
        return true;
    }
    *args = 0;

    for (int i = 0; i < MAX_COMMAND_PARAMETERS; i++) {
        parameters[i] = strtok_r(rest, FIELD_SEPERATOR, &rest);
        if (parameters[i] == NULL) {
            break;
        }
        (*args)++;
    }
    printf("Command %s parsed with %d arguments\n", *command, *args);

    return false;
}

void main() {

    char *cmd;
    char *par[MAX_COMMAND_PARAMETERS];
    uint8_t args;

    char buffer[MAX_INPUT_CHARACTERS];

    while(true) {


        printf("Enter a command :");
        scanf("%100s", buffer);

        if (process_buffer(&cmd, par, &args, buffer)) {
            continue;
        }

        printf("Command parsed %s.\n", cmd);

        if (strncmp("help", cmd, MAX_COMMAND_CHARACTERS) == 0) {
    
            if(args != 0) {
                printf("Command %s wrong arguments\n", cmd);
                continue;
            }
            
            print_help_menu();
            printf("Command %s finished\n", cmd);
        
        } else if (strncmp("led", cmd, MAX_COMMAND_CHARACTERS) == 0) {
                        
            print_led_command(par, args);
            printf("Command %s finished\n", cmd);
        
        } else if (strncmp("button", cmd, MAX_COMMAND_CHARACTERS) == 0) {
            
            if(args != 0) {
                printf("Command %s wrong arguments\n", cmd);
                continue;
            }

            print_button_command();
            printf("Command %s finished\n", cmd);

        } else {
            printf("Command not found\n");
        }
    }
}

File: python/simple_menu.py
import random

MAX_COMMAND_CHARACTERS = 6
MAX_COMMAND_PARAMETERS = 5
FIELD_SEPARATOR = ","
MAX_INPUT_CHARACTERS = 100

class CommandProcessor:
    def __init__(self):
        self.cmd = ""
        self.par = []

    def print_help_menu(self):
        print("This is where a description of the commands would be")

    def print_button_command(self):
        pressed = random.choice([True, False])
        print(f"Button status: {'pressed' if pressed else 'Not pressed'}")

    def print_led_command(self):
        if not self.par:
            print("Led command wrong parameters")
            return

        if self.par[0] == "on":
            print("Led is on")
        elif self.par[0] == "off":
            print("Led is off")
        elif self.par[0] == "blink":
            if len(self.par) != 2:
                print("Led blinking command wrong parameters")
            else:
                print(f"Led is blinking with {self.par[1]} Hz")
        else:
            print("Led command wrong parameters")

    def process_buffer(self, buffer):
        if not buffer:
            return True

        parts = buffer.split(FIELD_SEPARATOR, 1)
        self.cmd = parts[0]
        self.par = parts[1].split(FIELD_SEPARATOR) if len(parts) > 1 else []

        print(f"Command {self.cmd} parsed with {len(self.par)} arguments")

        return False

def main():
    processor = CommandProcessor()

    while True:
        buffer = input("Enter a command: ")

        if processor.process_buffer(buffer):
            continue

        print(f"Command parsed {processor.cmd}.")

        if processor.cmd == "help":
            if len(processor.par) != 0:
                print(f"Command {processor.cmd} wrong arguments")
                continue

            processor.print_help_menu()
            print(f"Command {processor.cmd} finished")

        elif processor.cmd == "led":
            processor.print_led_command()
            print(f"Command {processor.cmd} finished")

        elif processor.cmd == "button":
            if len(processor.par) != 0:
                print(f"Command {processor.cmd} wrong arguments")
                continue

            processor.print_button_command()
            print(f"Command {processor.cmd} finished")

        else:
            print("Command not found")

if __name__ == "__main__":
    main()
